File: public/codes/timSort.py
```python
def tim_sort(array, set_array, set_comparing_indices, set_swapping_indices, set_sorted_indices, speed, is_sorting):
    n = len(array)
    MIN_MERGE = 32

    def insertion_sort_tim(arr, left, right):
        for i in range(left + 1, right + 1):
            temp = arr[i]
            j = i - 1
            while j >= left and arr[j] > temp:
                if not is_sorting.current: return
                set_comparing_indices([j, j + 1])
                set_array(list(arr))
                time.sleep(speed / 1000.0)
                arr[j + 1] = arr[j]
                j -= 1
            arr[j + 1] = temp
            set_array(list(arr))
            time.sleep(speed / 1000.0)

    def merge_tim(arr, l, m, r):
        len1 = m - l + 1
        len2 = r - m
        left_arr = [0] * len1
        right_arr = [0] * len2

        for i in range(len1): left_arr[i] = arr[l + i]
        for i in range(len2): right_arr[i] = arr[m + 1 + i]

        i = 0
        j = 0
        k = l

        while i < len1 and j < len2:
            if not is_sorting.current: return
            set_comparing_indices([l + i, m + 1 + j])
            set_array(list(arr))
            time.sleep(speed / 1000.0)
            if left_arr[i] <= right_arr[j]:
                arr[k] = left_arr[i]
                i += 1
            else:
                arr[k] = right_arr[j]
                j += 1
            set_array(list(arr))
            time.sleep(speed / 1000.0)
            k += 1

        while i < len1:
            if not is_sorting.current: return
            arr[k] = left_arr[i]
            set_array(list(arr))
            time.sleep(speed / 1000.0)
            k += 1
            i += 1

        while j < len2:
            if not is_sorting.current: return
            arr[k] = right_arr[j]
            set_array(list(arr))
            time.sleep(speed / 1000.0)
            k += 1
            j += 1

    for i in range(0, n, MIN_MERGE):
        insertion_sort_tim(array, i, min((i + MIN_MERGE - 1), (n - 1)))

    size = MIN_MERGE
    while size < n:
        left = 0
        while left < n:
            mid = left + size - 1
            right = min((left + 2 * size - 1), (n - 1))
            if mid < right:
                merge_tim(array, left, mid, right)
            left += 2 * size
        size *= 2

    set_sorted_indices(list(range(n)))
    set_comparing_indices([])

import time
```

File: public/codes/timSort.js
async function timSort(array, setArray, setComparingIndices, setSwappingIndices, setSortedIndices, speed, isSorting) {
    const n = array.length;
    const MIN_MERGE = 32;

    const insertionSortTim = async (arr, left, right) => {
        for (let i = left + 1; i <= right; i++) {
            let temp = arr[i];
            let j = i - 1;
            while (j >= left && arr[j] > temp) {
                if (!isSorting.current) return;
                setComparingIndices([j, j + 1]);
                setArray([...array]);
                await sleep(speed);
                arr[j + 1] = arr[j];
                j--;
            }
            arr[j + 1] = temp;
            setArray([...array]);
            await sleep(speed);
        }
    };

    const mergeTim = async (arr, l, m, r) => {
        const len1 = m - l + 1;
        const len2 = r - m;
        const leftArr = new Array(len1);
        const rightArr = new Array(len2);

        for (let i = 0; i < len1; i++) leftArr[i] = arr[l + i];
        for (let i = 0; i < len2; i++) rightArr[i] = arr[m + 1 + i];

        let i = 0;
        let j = 0;
        let k = l;

        while (i < len1 && j < len2) {
            if (!isSorting.current) return;
            setComparingIndices([l + i, m + 1 + j]);
            setArray([...array]);
            await sleep(speed);
            if (leftArr[i] <= rightArr[j]) {
                arr[k] = leftArr[i];
                i++;
            } else {
                arr[k] = rightArr[j];
                j++;
            }
            setArray([...arr]);
            await sleep(speed);
            k++;
        }

        while (i < len1) {
            if (!isSorting.current) return;
            arr[k] = leftArr[i];
            setArray([...arr]);
            await sleep(speed);
            k++;
            i++;
        }

        while (j < len2) {
            if (!isSorting.current) return;
            arr[k] = rightArr[j];
            setArray([...arr]);
            await sleep(speed);
            k++;
            j++;
        }
    };

    for (let i = 0; i < n; i += MIN_MERGE) {
        await insertionSortTim(array, i, Math.min((i + MIN_MERGE - 1), (n - 1)));
    }

    for (let size = MIN_MERGE; size < n; size = 2 * size) {
        for (let left = 0; left < n; left += 2 * size) {
            const mid = left + size - 1;
            const right = Math.min((left + 2 * size - 1), (n - 1));
            if (mid < right) {
                await mergeTim(array, left, mid, right);
            }
        }
    }

    setSortedIndices(Array.from({ length: n }, (_, i) => i));
    setComparingIndices([]);
}

// Helper function (assuming it's defined elsewhere or passed)
function sleep(ms) {
    return new Promise(resolve => setTimeout(resolve, ms));
}